src/calliope/config.py

```diff
@@ -0,0 +1,369 @@
+    """
+
+    horizon: str = Field(default="48h")
+    """
+    Operate mode rolling `horizon`, given as a pandas frequency string.
+    See [here](https://pandas.pydata.org/docs/user_guide/timeseries.html#offset-aliases) for a list of frequency aliases.
+    Must be ≥ `window`
+    """
+
+    use_cap_results: bool = Field(default=False)
+    """If the model already contains `plan` mode results, use those optimal capacities as input parameters to the `operate` mode run."""
+
+    start_window_idx: int = Field(default=0, repr=False, exclude=True)
+    """Which time window to build. This is used to track the window when re-building the model part way through solving in `operate` mode."""
+
+
+class Build(ModeBaseModel):
+    """Base configuration options used when building a Calliope optimisation problem (`calliope.Model.build`)."""
+
+    model_config = {
+        "title": "build",
+        "extra": "allow",
+        "revalidate_instances": "always",
+    }
+    add_math: UniqueList[str] = Field(default=[])
+    """
+    List of references to files which contain additional mathematical formulations to be applied on top of or instead of the base mode math.
+    If referring to an pre-defined Calliope math file (see documentation for available files), do not append the reference with ".yaml".
+    If referring to your own math file, ensure the file type is given as a suffix (".yaml" or ".yml").
+    Relative paths will be assumed to be relative to the model definition file given when creating a calliope Model (`calliope.Model(model_definition=...)`)
+    """
+
+    ignore_mode_math: bool = Field(default=False)
+    """
+    If True, do not initialise the mathematical formulation with the pre-defined math for the given run `mode`.
+    This option can be used to completely re-define the Calliope mathematical formulation.
+    """
+
+    backend: Literal["pyomo", "gurobi"] = Field(default="pyomo")
+    """Module with which to build the optimisation problem."""
+
+    ensure_feasibility: bool = Field(default=False)
+    """
+    Whether to include decision variables in the model which will meet unmet demand or consume unused supply in the model so that the optimisation solves successfully.
+    This should only be used as a debugging option (as any unmet demand/unused supply is a sign of improper model formulation).
+    """
+
+    objective: str = Field(default="min_cost_optimisation")
+    """Name of internal objective function to use, from those defined in the pre-defined math and any applied additional math."""
+
+    pre_validate_math_strings: bool = Field(default=True)
+    """
+    If true, the Calliope math definition will be scanned for parsing errors _before_ undertaking the much more expensive operation of building the optimisation problem.
+    You can switch this off (e.g., if you know there are no parsing errors) to reduce overall build time.
+    """
+
+    operate: BuildOperate = BuildOperate()
+
```

Comment:
It hides it in the repr, try simply `model.config.build` which shows you all the config information but without the mode-based info. 

I did try the extension approach but didn't find a satisfactory solution that worked from the instatiation phase onwards. A few things that are _requirements_:

1. You can load build/solve config at the init stage (from YAML) and it is validated at this point (because you want you model to fail fast: `solve` config errors shouldn't crop up _after_ you've waited for a large model to build)
2. you can load e.g. `operate` mode config but not actually use it initially, because e.g. you start by running in `plan` mode

Given those two requirements, you have to be prepared to expect out-of-mode config straight away but if the config is non-extensible (to avoid adding spurious config at the `solve` stage, particularly) then validation will fail as it sees config options it doesn't expect and cannot allow. This tries to walk that tightrope by allowing the config options but hiding it from the user unless they explicitly are using it in their run. If you have a way to make it work as you suggest, I'd gladly update to go that way!

Reply:
Ok, in that case, I'll just treat calliope modes as non-user extensible too, which is fine.